### PR TITLE
WIP:If the user has an expired github OAuth token, re-authenticate them.

### LINKED ISSUE #115: Revoking Github Access prevents regranting access
screen shot 2018-04-18 at 10 05 16 am

As a user, if you click the `Revoke` button on the https://github.com/settings/applications page and they try to run a `/deploy` command in slack, you get the following error notification:

```Oops! We had a problem running your command, but we've been notified```

The correlated error found via rollbar is:

https://rollbar.com/Remind/slashdeploy/items/18/

```
#18 Octokit::Unauthorized: GET https://api.github.com/repos/remind101/stacks/deployments?sha=1: 401 - Bad credentials // See: https://developer.github.com/v3
```

This is likely because the cached oauth token in the SlashDeploy database is no longer valid.

**What happens**

It throws an exception.

**What I expect to happen**

I expect slashdeploy to catch this error and try once to re-add the SlashDeploy app to the users `Authorized OAuth Apps` thus generating and persisting a new token.

### PR DESCRIPTION
**Pull request by russellballestrini**
Fixes #115 

	modified:   lib/slashdeploy/auth.rb

I'm having trouble verifying this fix functionally. When I try I don't get any relavent error messages in my development env. I've manually disabled rollbar in dev but still nothing.

### CODE REVIEW

File: lib/slashdeploy/auth.rb [hunk 1]
```diff
@@ -31,11 +31,17 @@ def call(env)
 
       begin
         handler.call(env)
-      rescue User::MissingGitHubAccount
-        # If we don't know this slack user, we'll ask them to authenticate
-        # with GitHub. We encode and sign the Slack user id within the state
-        # param so we know what slack user they are when the hit the GitHub
-        # callback.
+      rescue User::MissingGitHubAccount, Octokit::Unauthorized
```
Comment (ejholmes) [inline lib/slashdeploy/auth.rb]: Doing this so broadly may end up having some unintended side effects. For example, if the access token is valid, but you don't have access to a specific repository, this is the exception that's raised.

We should probably have a way to "test" the token first, to verify that it's actually invalid, probably using the [`GET /user` api](https://developer.github.com/v3/users/#get-the-authenticated-user). That will let us know 100% that the token is invalid if we get a 401:

```console
$ curl https://api.github.com/user -i
HTTP/1.1 401 Unauthorized
Date: Thu, 19 Apr 2018 01:24:35 GMT
Content-Type: application/json; charset=utf-8
Content-Length: 135
Server: GitHub.com
Status: 401 Unauthorized

{
  "message": "Requires authentication",
  "documentation_url": "https://developer.github.com/v3/users/#get-the-authenticated-user"
}
```

Comment (russellballestrini) [reply lib/slashdeploy/auth.rb]: Octokit raises a different exception when access to the repo is unauthorized. Catching this exception is catching anything related to `Authentication` (passwords, tokens, etc) and it is an `HTTP 401`.

Any errors related to `Authorization` (like being authenticated but not allowed to view a particular repo) will throw an `HTTP 403`.

Reference: https://github.com/octokit/octokit.rb/blob/master/lib/octokit/error.rb#L238

**Comment by russellballestrini** [comment]
I'm not sure where to put the test.

File: lib/slashdeploy/auth.rb (hunk 1 above)
Comment (fergyfresh) [reply lib/slashdeploy/auth.rb]: I read through every spec and this seems like the best place for it, but I'm very new to this repo.

I was thinking maybe you could put the test here.
https://github.com/remind101/slashdeploy/blob/master/spec/controllers/sessions_controller_spec.rb

**Pull request merged**